### PR TITLE
add cpu reservation for 64 core hosts

### PR DESCRIPTION
**Pull request by twaldrop**

### CODE REVIEW

File: roles/nova-common/templates/etc/nova/nova.conf [hunk 1]
```diff
@@ -30,12 +30,18 @@ cpu_allocation_ratio={{ nova.cpu_allocation_ratio }}
...
+#Reserve CPUs on H8DGU or 64 core hosts
 {% if ansible_product_name == 'H8DGU' and ansible_processor_vcpus == 32 and inventory_hostname in groups['controller'] -%}
 vcpu_pin_set = "2-15,18-31"
 {% elif ansible_product_name == 'H8DGU' and ansible_processor_vcpus == 32 -%}
 vcpu_pin_set = "1-31"
+{% elif ansible_processor_vcpus == 64 and inventory_hostname in groups['controller'] -%}
+vcpu_pin_set = "2-31,33-63"
+{% elif ansible_processor_vcpus == 64 -%}
+vcpu_pin_set = "1-63"
 {% endif -%}
```
Comment (dlundquist) [inline roles/nova-common/templates/etc/nova/nova.conf]: So existing code reserves two cores (one [https://en.wikipedia.org/wiki/Bulldozer_(microarchitecture)#CMT CMT unit]) on each socket for this HW platform. If we are going to reserve cores on a wide range of HW platforms, we should not try such optimizations and use a formula to reserve 1 CPU core on non-controller nodes, and 4 CPU cores on controller nodes.

File: roles/nova-common/templates/etc/nova/nova.conf [hunk 2]
```diff
@@ -30,10 +30,12 @@ cpu_allocation_ratio={{ nova.cpu_allocation_ratio }}
 ram_allocation_ratio={{ nova.ram_allocation_ratio }}
 {% endif -%}
 
-{% if ansible_product_name == 'H8DGU' and ansible_processor_vcpus == 32 and inventory_hostname in groups['controller'] -%}
-vcpu_pin_set = "2-15,18-31"
-{% elif ansible_product_name == 'H8DGU' and ansible_processor_vcpus == 32 -%}
-vcpu_pin_set = "1-31"
+#Reserve CPUs on converged controllers and compute nodes
+{% set reserved_cores = if inventory_hostname in groups['controller'] is 4 else 1 %}
```
Comment (j2sol) [inline roles/nova-common/templates/etc/nova/nova.conf]: Is this missing a variable for the value?

**Comment by j2sol** [comment]
Looks right to me, and clean. Okay to merge once tests pass. :+1:

**Comment by j2sol** [comment]
Please rebase on master to pick up required fixes to pass tests.

**Comment by twaldrop** [comment]
rebased

**Comment by j2sol** [comment]
retest

**Comment by j2sol** [comment]
Looks like the syntax still isn't right. Are you testing locally? Do we have an Ansible version mismatch issue?

File: roles/nova-common/templates/etc/nova/nova.conf [hunk 3]
```diff
@@ -30,10 +30,12 @@ cpu_allocation_ratio={{ nova.cpu_allocation_ratio }}
...
-vcpu_pin_set = "2-15,18-31"
-{% elif ansible_product_name == 'H8DGU' and ansible_processor_vcpus == 32 -%}
-vcpu_pin_set = "1-31"
+#Reserve CPUs for host OS on converged controllers and compute hosts
+{% set reserved_cores = 4 if inventory_hostname in groups['controller'] else set reserved_cores = 1 %}
+{% set cpulimit = cpulimit | default(ansible_processor_vcpus) %}
+{% set cpu_core_count = cpulimit if cpulimit < ansible_processor_vcpus else set cpu_core_count = ansible_processor_vcpus %}
+{% if cpu_core_count > reserved_cores %}
+vcpu_pin_set = "{{ reserved_cores }}-{{ cpu_core_count - 1 }}"
 {% endif -%}
```
Comment (j2sol) [inline roles/nova-common/templates/etc/nova/nova.conf]: Give this a try:
```
{% set reserved_cores = 4 if inventory_hostname in groups['controller'] else 1 %}
{% set cpulimit = cpulimit | default(ansible_processor_vcpus) %}
{% set cpu_core_count = cpulimit if cpulimit < ansible_processor_vcpus else ansible_processor_vcpus %}
{% if cpu_core_ount > reserved_cores %}
vpu_pin_set = "{{ reserved_cores }}-{{ cpu_core_count - 1 }}"
{% endif -%}
```

**Pull request merged**